core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala

```diff
@@ -0,0 +1,135 @@
+  /**
+   * The replication in StorageLevel may be out of date. E.g., when the first block is added, the
+   * replication is 1. But when the second block with the same ID is added, the replication should
+   * become 2. To avoid scanning "blockManagers" to modify the replication number, we maintain
+   * "blockLocations" to get the replication quickly.
+   */
+  private val blockLocations = new mutable.HashMap[BlockId, mutable.ArrayBuffer[String]]
+
+  override def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit = {
+    val blockId = blockUpdated.blockUpdatedInfo.blockId
+    if (!blockId.isInstanceOf[StreamBlockId]) {
+      // Now we only monitor StreamBlocks
+      return
+    }
+    val blockManagerId = blockUpdated.blockUpdatedInfo.blockManagerId
+    val storageLevel = blockUpdated.blockUpdatedInfo.storageLevel
+    val memSize = blockUpdated.blockUpdatedInfo.memSize
+    val diskSize = blockUpdated.blockUpdatedInfo.diskSize
+    val externalBlockStoreSize = blockUpdated.blockUpdatedInfo.externalBlockStoreSize
+
+    synchronized {
+      // Drop the update info if the block manager is not registered
+      blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+        if (storageLevel.isValid) {
+          blocksInBlockManager.put(blockId,
+            BlockUIData(
+              blockId,
+              blockManagerId.hostPort,
+              storageLevel,
+              memSize,
+              diskSize,
+              externalBlockStoreSize)
+          )
+          val locations = blockLocations.getOrElseUpdate(blockId, new mutable.ArrayBuffer[String])
+          locations += blockManagerId.hostPort
+        } else {
+          // If isValid is not true, it means we should drop the block.
+          blocksInBlockManager -= blockId
+          removeLocationFromBlockLocations(blockId, blockManagerId.hostPort)
+        }
+      }
+    }
+  }
+
+  override def onBlockManagerAdded(blockManagerAdded: SparkListenerBlockManagerAdded): Unit = {
+    synchronized {
+      blockManagers.put(blockManagerAdded.blockManagerId, mutable.HashMap())
+    }
+  }
+
+  private def removeLocationFromBlockLocations(blockId: BlockId, location: String): Unit = {
+    synchronized {
+      blockLocations.get(blockId).foreach { locations =>
+        locations -= location
+        if (locations.isEmpty) {
+          blockLocations -= blockId
+        }
+      }
+    }
+  }
+
+  override def onBlockManagerRemoved(
+      blockManagerRemoved: SparkListenerBlockManagerRemoved): Unit = synchronized {
+    val blockManagerId = blockManagerRemoved.blockManagerId
+    blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+      blocksInBlockManager.keys.foreach(
+        removeLocationFromBlockLocations(_, blockManagerId.hostPort))
+    }
+    blockManagers -= blockManagerId
+  }
+
+  def allExecutorStreamBlockStatus: Seq[ExecutorStreamBlockStatus] = synchronized {
```

Comment:
These public functions can get called at different times leading to inconsistent values in the UI. Can you define a method that returns both information in a single synchronized call?